### ISSUE TITLE
[question] Conan 2.x: updating a dependency in multiple lockfiles

### ISSUE THREAD

**Issue by tbsuht** [conan-io/conan#13609]
### What is your question?

Hi,

I'm investigating whether 2.0 lockfiles are a feasible solution for our release workflow, where the basic situation looks like this:

We have a bunch of projects, each having its own conanfile, which will combined build a releasable product.
Cyclic before a release we start to stabilize, which seems like a good point to create lockfiles for the roots (applications) of our dependency graphs.

Whats the best way to incorporate changes (bugfixes) into the lockfiles after creating them? This is e.g. the case if some lib in a dependency graph changes.

I've watched the presentation videos of Conan 2.0, so I think `cd lib-that-change && conan create . --version=... -l old-lockfile --lockfile-out new-lockfile` should be the way to go.

Now my question is:
I have created a lot of lockfiles initially, should I just parse where my changed lib is part of and call the mentioned command to update them? That seems kind of redundant as we will trigger a lot of builds. I know that there is `conan lock add` but I don't like that this command is basically a raw manipulation of the file stated in the docs without checks.

### Have you read the CONTRIBUTING guide?

- [x] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @tbsuht 

This is a very good question, and it is something that is pending of some docs to better explain it.

The core idea for an escalable, paralelizable pipeline is something like this:
- You have a bunch of ``conanfile.py`` representing your final applications or products, we call them the "products".
- When you install the dependencies and build one of those products you can capture the lockfile of that particular product. You can do that for every product, so you will have as many lockfiles as products.
- Then, some random package, that might be used or not in some, none or all products, gets a fix, and lets say gets a new version, from ``pkga/1.0.0`` to ``pkga/1.0.1``
- The first part of the pipeline is the "package" pipeline/CI where ``pkga/1.0.1`` is tested, for typical configurations, and the typical unit/integration/functional testing of ``pkga/1.0.1`` happens. From there you basically get a new reference, with its revision like ``pkga/1.0.1#revision1``
- For every product lockfile you can ``conan lock add --requires=pkga/1.0.1#revision --lockfile=app1.lock --lockfile-out=app1_new.lock``
- With that lockfile you can compute what needs to be built with ``conan graph build-order --requires=product1/1.0 --lockfile=app1_new.lock --format=json > product1_build_order.json``. This will return a json with what needs to be built in what order. If the new version ``pkga/1.0.1`` is not used, then nothing will need to be built.
- Merge all the productX_build_order.json with ``conan graph build-order-merge`` into a single "build-order"
- Iterate the resulting aggregated build-order to build what needs to be built

This guarantees that no package is unnecessarily rebuilt more than once like:
- product1 -> pkgc/1.0 -> pkgb/1.0 ->pkga/1.0.1
- product2 -> pkgb/1.0 -> pkga/1.0.1

In this case, only 1 build of ``pkgb/1.0`` against the new dependency should be necessary, and this can be done just once if using the aggregated build order.

Of course there are ways to simplify the above pipeline, I am describing the most advanced case, when there are multiple products, and we want to do a decentralized, parallel (the ``build-order`` returns a list of lists with packages that can be built in parallel), efficient (the ``package_id`` knows what needs to be built and what not)

Note that some products that might have ``pkg/1.0.1`` outside of the version range, for example if some is pinning a ``requires = "pkga/1.0.0"``, that product will stay with the older ``pkga`` version, only the products that ``requires = "pkga/[>=1.0.0 <2.0]"`` will use the new version. This is good, and a way to control breaking changes in a nice way.

**Comment by tbsuht** [comment]
Hi @memsharded,

thanks for the detailed example, this looks promising!

> For every product lockfile you can conan lock add --requires=pkga/1.0.1#revision --lockfile=app1.lock --lockfile-out=app1_new.lock

Is it possible to really replace a version and not just add it?

> With that lockfile you can compute what needs to be built with conan graph build-order --requires=product1/1.0 --lockfile=app1_new.lock --format=json > product1_build_order.json. This will return a json with what needs to be built in what order. If the new version pkga/1.0.1 is not used, then nothing will need to be built.

This was not working for me.
```
$ conan graph build-order --requires=app/1.0.0 --lockfile=updated-conan.lock --format=json > abc.json
Using lockfile: '/workspaces/dev/conan-lockfiles/app/updated-conan.lock'

======== Computing dependency graph ========
ERROR: Requirement 'app/1.0.0' not in lockfile
```
I changed that to `conan graph build-order . --lockfile updated-conan.lock --format=json > order.json`. Is it possible to provide here a reference instead of the path to the conanfile?

> Merge all the productX_build_order.json with conan graph build-order-merge into a single "build-order"

Does it make sense in general to only work with the merged lockfile? This is easier to handle and store. This would then represent all "products" in one file.

It would be also nice to have some sanity checks for the merge in our case: e.g. we only want to have a single version of e.g. a lib in the merged lockfile / across the product lockfiles. This needs to be implemented by our own right?

**Comment by memsharded** [comment]
> Is it possible to really replace a version and not just add it?

The idea is that it shouldn't be necessary to replace it. When the new version satisfies a version range, it will be used, and if the new version does not satisfy a version range, removing the previous existing old version will actually break things. It is not possible to force via lockfiles to override versions beyond the defined version ranges.

> This was not working for me.
> I changed that to conan graph build-order . --lockfile updated-conan.lock --format=json > order.json. Is it possible to provide here a reference instead of the path to the conanfile?

For this to work, ``app/1.0.0`` should have been packaged, and created too, and then capture its lockfile with ``conan lock create --requires=xxx``, and not ``conan lock create <path>`` otherwise the ``<path>`` part will not be fully locked, as it is not a package.


> Does it make sense in general to only work with the merged lockfile? This is easier to handle and store. This would then represent all "products" in one file.
> It would be also nice to have some sanity checks for the merge in our case: e.g. we only want to have a single version of e.g. a lib in the merged lockfile / across the product lockfiles. This needs to be implemented by our own right?

That depends. If you really want to enforce and try to converge as much as possible to use the same versions across different products, then it might be possible to use the same lockfile for all of them, but note that this is not strictly necessary per se, as each product can have its own, separate dependency graph with its own "snapshot" (lockfile of it).

**Comment by Arunsid** [comment]
> Hi @tbsuht
> 
> This is a very good question, and it is something that is pending of some docs to better explain it.
> 
> The core idea for an escalable, paralelizable pipeline is something like this:
> 
> * You have a bunch of `conanfile.py` representing your final applications or products, we call them the "products".
> * When you install the dependencies and build one of those products you can capture the lockfile of that particular product. You can do that for every product, so you will have as many lockfiles as products.
> * Then, some random package, that might be used or not in some, none or all products, gets a fix, and lets say gets a new version, from `pkga/1.0.0` to `pkga/1.0.1`
> * The first part of the pipeline is the "package" pipeline/CI where `pkga/1.0.1` is tested, for typical configurations, and the typical unit/integration/functional testing of `pkga/1.0.1` happens. From there you basically get a new reference, with its revision like `pkga/1.0.1#revision1`
> * For every product lockfile you can `conan lock add --requires=pkga/1.0.1#revision --lockfile=app1.lock --lockfile-out=app1_new.lock`
> * With that lockfile you can compute what needs to be built with `conan graph build-order --requires=product1/1.0 --lockfile=app1_new.lock --format=json > product1_build_order.json`. This will return a json with what needs to be built in what order. If the new version `pkga/1.0.1` is not used, then nothing will need to be built.
> * Merge all the productX_build_order.json with `conan graph build-order-merge` into a single "build-order"
> * Iterate the resulting aggregated build-order to build what needs to be built
> 
> This guarantees that no package is unnecessarily rebuilt more than once like:
> 
> * product1 -> pkgc/1.0 -> pkgb/1.0 ->pkga/1.0.1
> * product2 -> pkgb/1.0 -> pkga/1.0.1
> 
> In this case, only 1 build of `pkgb/1.0` against the new dependency should be necessary, and this can be done just once if using the aggregated build order.
> 
> Of course there are ways to simplify the above pipeline, I am describing the most advanced case, when there are multiple products, and we want to do a decentralized, parallel (the `build-order` returns a list of lists with packages that can be built in parallel), efficient (the `package_id` knows what needs to be built and what not)
> 
> Note that some products that might have `pkg/1.0.1` outside of the version range, for example if some is pinning a `requires = "pkga/1.0.0"`, that product will stay with the older `pkga` version, only the products that `requires = "pkga/[>=1.0.0 <2.0]"` will use the new version. This is good, and a way to control breaking changes in a nice way.

Regarding the same case, how will be the solution in conan 1.x versions ?Thanks

**Comment by memsharded** [comment]
@Arunsid I know that this is probably not the answer you are expecting/looking for, but here are my 2 cents: It would be more challenging to try to set-up and implement this in Conan 1.X lockfiles than moving to Conan 2. And as the migration to Conan 2 needs to happen anyway (Conan 2 is already 19 months old, already in 2.7, ConanCenter will soon freeze the updates for Conan 1.X packages), then I'd clearly prioritize the upgrade to Conan 2, and then implement the CI/lockfiles flow from there.
